dspy/utils/caching.py

```diff
@@ -12,3 +18,58 @@ def create_subdir_in_cachedir(subdir: str) -> str:
     subdir = os.path.abspath(subdir)
     os.makedirs(subdir, exist_ok=True)
     return subdir
+
+
+class LRUCache(OrderedDict):
```

Comment:
@bahtman Can we extend some of the testing in https://github.com/stanfordnlp/dspy/blob/main/tests/caching/test_caching.py to verify that in-memory caching works properly?

Reply:
@dbczumar I had this realization today — any reason why we're not just using joblib? I think it was because it was in the dsp folder (and we're trying to move away from that), but joblib seemed like it handled a lot of things well out of the box.

Reply:
@CyrusNuevoDia What do you think was handled by joblib? Joblib handled everything strictly worse than a basic DiskCache. In 2.4 and earlier, we combined joblib with manual lru_cache.

Reply:
@okhat oh thanks, I wasn't privy to that context. Makes sense!

I liked how it automatically wrote/read from disk — maybe this LRUCache + DiskCache is the way to go then?

Reply:
One issue I hadn't accounted for here is thread safety. This cache is likely to be called in parallel during optimization or evaluation. The implementation needs to be thread safe. I suspect relying on a threadsafe cache (e.g. continuing to use functools) is probably best. @CyrusNuevoDia thoughts on removing the caching portions from this PR and getting the streaming functionality merged independently?

Reply:
Fixed by merging in your cache